app/controllers/packages_controller.rb

```diff
@@ -0,0 +1,78 @@
+class PackagesController < ApplicationController
+  skip_before_filter :verify_authenticity_token, only: [:save]
+
+  def estimate_request # api endpoint
+      @betsy_shipping = params
+      origin
+      destination
+      package
+
+      begin
+        ups = ActiveShipping::UPS.new(:login => ENV["ACTIVESHIPPING_UPS_LOGIN"],
+                                      :password => ENV["ACTIVESHIPPING_UPS_PASSWORD"],
+                                      :key => ENV["ACTIVESHIPPING_UPS_KEY"])
+        ups_response = ups.find_rates(@origin, @destination, @package)
+        ups_rates = ups_response.rates.sort_by(&:price).collect { |rate| [rate.service_name, rate.price, rate.delivery_date] }
+        usps = ActiveShipping::USPS.new(:login => ENV["ACTIVESHIPPING_USPS_LOGIN"])
+        usps_response = usps.find_rates(@origin, @destination, @package)
+        usps_rates = usps_response.rates.sort_by(&:price).collect { |rate| [rate.service_name, rate.price] }
+
+        carrier_rates = []
+        carrier_rates.push(ups_rates, usps_rates)
+        render json: carrier_rates.as_json
+      rescue ActiveShipping::ResponseError => error
+       error_json = { message: error.message }
+       render json: error_json.as_json, status: 400
+      end
+  end
+
+  def save
+    service_array = params["order"]["estimate"]["service"].split(" ")
+    price = service_array.pop.to_f
+    service_type = service_array.join(" ")
+    weight = params["order"]["estimate"]["service_info"]["estimate"]["total_weight"].to_i
+    @box_size = params["order"]["estimate"]["service_info"]["estimate"]["box_size"]
```

Comment:
This is a really complicated interface. My first reaction is that you should document the API so that people will know what parameters to provide but then I realized that the better solution is to simplify the API parameters. Why do you have three or four levels deep of parameters? It would be much more _expected_ to match the Rails restful interface for a controller and just provide the parameters as `params[:package]` with weight, sizing, order_id, price, and service_type parameters.

Especially confusing to me is the first three lines where a parameter calls `:service` is being used to also include a price, that is space-separated. String parsing is the most common place for security vulnerabilities so you should avoid it when you can.